lib/ClangImporter/DWARFImporter.cpp

```diff
@@ -0,0 +1,178 @@
+  }
+
+  virtual StringRef getFilename() const override { return ""; }
+
+  virtual const clang::Module *getUnderlyingClangModule() const override {
+    return nullptr;
+  }
+
+  static bool classof(const FileUnit *file) {
+    return file->getKind() == FileUnitKind::DWARFModule;
+  }
+  static bool classof(const DeclContext *DC) {
+    return isa<FileUnit>(DC) && classof(cast<FileUnit>(DC));
+  }
+};
+
+ModuleDecl *ClangImporter::Implementation::loadModuleDWARF(
+    SourceLoc importLoc, ArrayRef<std::pair<Identifier, SourceLoc>> path) {
+  // There's no importing from debug info if no importer is installed.
+  if (!DWARFImporter)
+    return nullptr;
+
+  // FIXME: Implement submodule support!
+  Identifier name = path[0].first;
+  auto it = DWARFModuleUnits.find(name);
+  if (it != DWARFModuleUnits.end())
+    return it->second->getParentModule();
+
+  auto *decl = ModuleDecl::create(name, SwiftContext);
+  decl->setIsNonSwiftModule();
+  decl->setHasResolvedImports();
+  auto wrapperUnit = new (SwiftContext) DWARFModuleUnit(*decl, *this);
+  DWARFModuleUnits.insert({name, wrapperUnit});
+  decl->addFile(*wrapperUnit);
+
+  // Force load overlay modules for all imported modules.
+  decl->forAllVisibleModules({}, [](ModuleDecl::ImportedModule import) {});
+
+  // Register the module with the ASTContext so it is available for lookups.
+  ModuleDecl *&loaded = SwiftContext.LoadedModules[name];
+  if (!loaded)
+    loaded = decl;
+
+  return decl;
+}
+
+void ClangImporter::Implementation::lookupValueDWARF(
+    ModuleDecl::AccessPathTy accessPath, DeclName name, NLKind lookupKind,
+    SmallVectorImpl<ValueDecl *> &results) {
+  if (!swift::ModuleDecl::matchesAccessPath(accessPath, name))
+    return;
+
+  if (lookupKind != NLKind::QualifiedLookup)
+    return;
+
+  if (!DWARFImporter)
+    return;
+
+  SmallVector<clang::Decl *, 4> decls;
+  DWARFImporter->lookupValue(name.getBaseIdentifier().str(), llvm::None, decls);
+  for (auto *clangDecl : decls) {
+    auto *namedDecl = dyn_cast_or_null<clang::NamedDecl>(clangDecl);
+    if (!namedDecl)
+      continue;
+    auto *swiftDecl = cast_or_null<ValueDecl>(
+        importDeclReal(namedDecl->getMostRecentDecl(), CurrentVersion));
+    if (!swiftDecl)
+      continue;
+
+    if (swiftDecl->getFullName().matchesRef(name) &&
+        swiftDecl->getDeclContext()->isModuleScopeContext())
+      results.push_back(swiftDecl);
+  }
+}
+
+void ClangImporter::Implementation::lookupTypeDeclDWARF(
+    StringRef rawName, ClangTypeKind kind,
+    llvm::function_ref<void(TypeDecl *)> receiver) {
+  if (!DWARFImporter)
+    return;
+
+  SmallVector<clang::Decl *, 1> decls;
+  DWARFImporter->lookupValue(rawName, kind, decls);
+  for (auto *clangDecl : decls) {
+    if (!isa<clang::TypeDecl>(clangDecl) &&
+        !isa<clang::ObjCContainerDecl>(clangDecl) &&
+        !isa<clang::ObjCCompatibleAliasDecl>(clangDecl)) {
+      continue;
+    }
+    auto *namedDecl = dyn_cast_or_null<clang::NamedDecl>(clangDecl);
```

Comment:
This one should just be `cast`, no? Since we've already tested for specific subclasses of `NamedDecl` above (and assumed it's non-null).